deepmd/train/trainer.py

```diff
@@ -539,56 +532,81 @@ def train (self,
-                # again the data object knows appropriate test data shape,
-                # there is no need to slice again!
-                # feed_dict_test[self.place_holders[kk]] = np.reshape(test_data[kk][:self.numb_test[data.pick_idx]], [-1])
-                feed_dict_test[self.place_holders[kk]] = np.reshape(test_data[kk], [-1])
-        for ii in ['type'] :
-            feed_dict_test[self.place_holders[ii]] = np.reshape(test_data[ii], [-1])            
-        for ii in ['natoms_vec', 'default_mesh'] :
-            feed_dict_test[self.place_holders[ii]] = test_data[ii]
-        feed_dict_test[self.place_holders['is_training']] = False
+                         train_data,
+                         valid_data,
+                         print_header=False):
+        train_results = self.get_evaluation_results(train_data, self.valid_numb_batch)
+        valid_results = self.get_evaluation_results(valid_data, self.valid_numb_batch)
 
         cur_batch = self.cur_batch
         current_lr = self.sess.run(self.learning_rate)
-        if self.run_opt.is_chief:
-            print_str = "%7d" % cur_batch
-            print_str += self.loss.print_on_training(
-                tb_writer,
-                cur_batch,
-                self.sess,
-                test_data['natoms_vec'],
-                feed_dict_test,
-                feed_dict_batch
-            )
-
-            print_str += "   %8.1e\n" % current_lr
-            fp.write(print_str)
-            fp.flush ()
+        if print_header:
+            self.print_header(fp, valid_results)
+        self.print_on_training(fp, train_results, valid_results, cur_batch, current_lr)
+
+    @staticmethod
+    def print_header(fp, results):
+        print_str = ''
+        print_str += "# %5s" % 'batch'
+        prop_fmt =  '   %11s %11s'
+        for k in results.keys():
+            print_str += prop_fmt % (k + '_val', k + '_trn')
+        print_str += '   %8s\n' % 'lr'
+        fp.write(print_str)
+        fp.flush()
+
+    @staticmethod
+    def print_on_training(fp, train_results, valid_results, cur_batch, cur_lr):
+        print_str = ''
+        print_str += "%7d" % cur_batch
+        prop_fmt = "   %11.2e %11.2e"
+        for k in valid_results.keys():
+            # assert k in train_results.keys()
+            print_str += prop_fmt % (valid_results[k], train_results[k])
+        print_str += "   %8.1e\n" % cur_lr
+        fp.write(print_str)
+        fp.flush()
+
+    def get_evaluation_results(self, data, numb_batch):
+        sum_results = None
+        for i in range(numb_batch):
+            feed_dict, natoms = self.get_feed_dict(data, is_training=False)
+            results = self.loss.eval(self.sess, feed_dict, natoms)
+            if sum_results is None:
+                sum_results = results
+            else:
+                for k, v in results:
+                    sum_results[k] += v
+        avg_results = {k: v / numb_batch for k, v in sum_results.items()}
```

Comment:
Here the error should be calculate in the sense of root mean square, rather than simply mean.